algorithms/linfa-clustering/src/k_means/init.rs

```diff
@@ -0,0 +1,389 @@
+    pub(crate) fn run(
+        &self,
+        n_clusters: usize,
+        observations: &'a ArrayView2<'a, F>,
+        rng: &mut impl Rng,
+    ) -> Array2<F> {
+        match self {
+            Self::Random => random_init(n_clusters, observations, rng),
+            Self::KMeansPlusPlus => k_means_plusplus(n_clusters, observations, rng),
+            Self::KMeansPara(rng_func) => k_means_para(n_clusters, observations, rng, rng_func),
+        }
+    }
+}
+
+/// Pick random points from the input matrix as centroids
+fn random_init<F: Float>(
+    n_clusters: usize,
+    observations: &ArrayView2<F>,
+    rng: &mut impl Rng,
+) -> Array2<F> {
+    let (n_samples, _) = observations.dim();
+    let indices = rand::seq::index::sample(rng, n_samples, n_clusters).into_vec();
+    observations.select(Axis(0), &indices)
+}
+
+/// Selects centroids using the KMeans++ initialization algorithm. The weights determine the
+/// likeliness of an input point to be selected as a centroid relative to other points. The higher
+/// the weight, the more likely the point will be selected as a centroid.
+fn weighted_k_means_plusplus<F: Float + SampleUniform + for<'a> AddAssign<&'a F>>(
+    n_clusters: usize,
+    observations: &ArrayView2<F>,
+    weights: &ArrayView1<F>,
+    rng: &mut impl Rng,
+) -> Array2<F> {
+    let (n_samples, n_features) = observations.dim();
+    assert_eq!(n_samples, weights.len());
+    assert_ne!(weights.sum(), F::zero());
+
+    let mut centroids = Array2::zeros((n_clusters, n_features));
+    // Select 1st centroid from the input randomly purely based on the weights.
+    let first_idx = WeightedIndex::new(weights.iter())
+        .expect("invalid weights")
+        .sample(rng);
+    centroids.row_mut(0).assign(&observations.row(first_idx));
+
+    let mut dists = Array1::zeros(n_samples);
+    for c_cnt in 1..n_clusters {
+        update_min_dists(&centroids.slice(s![0..c_cnt, ..]), observations, &mut dists);
+        dists *= weights;
+
+        // The probability of a point being selected as the next centroid is proportional to its
+        // distance from its closest centroid multiplied by its weight.
+        let centroid_idx = WeightedIndex::new(dists.iter())
+            .map(|idx| idx.sample(rng))
+            // This only errs if all of dists is 0, which means every point is assigned to a
+            // centroid, so extra centroids don't matter and can be any index.
+            .unwrap_or(0);
+        centroids
+            .row_mut(c_cnt)
+            .assign(&observations.row(centroid_idx));
+    }
+    centroids
+}
+
+/// KMeans++ initialization algorithm without biased weights
+fn k_means_plusplus<F: Float + SampleUniform + for<'a> AddAssign<&'a F>>(
+    n_clusters: usize,
+    observations: &ArrayView2<F>,
+    rng: &mut impl Rng,
+) -> Array2<F> {
+    weighted_k_means_plusplus(
+        n_clusters,
+        observations,
+        &Array1::ones(observations.nrows()).view(),
```

Comment:
`DatasetBase` hardcodes its weights to f32, which means I can't multiply it directly with the observations, which are of the type F. Also using datasets doesn't make the code any simpler, since the algorithm is still the same.